### PR TITLE
Fix BoostedDoubleSVProd crashing with nan/inf daughters

### PR DESCRIPTION
**Pull request by andrzejnovak**
#### PR description:

Fix for a problem reported in https://its.cern.ch/jira/browse/CMSCOMPPR-6445. Crashes caused by daughters with nan/inf values. Protection is added.

#### PR validation:

Local test, minor straightforward change

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoBTag/SecondaryVertex/plugins/BoostedDoubleSVProducer.cc [hunk 1]
```diff
@@ -655,8 +655,13 @@ void BoostedDoubleSVProducer::calcNsubjettiness(const reco::JetBaseRef& jet,
             edm::LogWarning("MissingJetConstituent")
                 << "Jet constituent required for N-subjettiness computation is missing!";
         }
-      } else
+      } else {
+        // Check if any values were nan or inf
+        float valcheck = daughter->px() + daughter->py() + daughter->pz() + daughter->energy();
+        if (std::isnan(valcheck) || std::isinf(valcheck))
```
Comment (perrotta) [inline RecoBTag/SecondaryVertex/plugins/BoostedDoubleSVProducer.cc]: As from the warning message displayed by the static analyzer:
"std::isnan / std::isinf does not work when fast-math is used. Please use edm::isNotFinite from 'FWCore/Utilities/interface/isFinite.h'"

**Comment by perrotta** [comment]
please test

**Comment by mverzett** [comment]
@fabiocos given this fix prevents the crash in some NanoAOD processes we would like to backport this to 10_6 and enter the same patch release as the DeepTau PR.
Thanks.

**Comment by fabiocos** [comment]
@mverzett so please prepare the backport for next patch

**Comment by perrotta** [comment]
@andrzejnovak @mverzett : besides putting a protection here, did you ever investigated where those nan's come out, and whether it is possible to fix them already when they are produced?

**Comment by perrotta** [comment]
Once the origin of the nan's is identificated, then if this "fix" is maintained it must be correlated with either a LogWarning or LogError, depending on the severity and frequency ecpected for the issue

**Comment by andrzejnovak** [comment]
I don't understand what is blocking this. Will you merge it if I add the warning/log. Or do you refuse to add a temp fix until the origin of the nans and infs is found?

**Comment by perrotta** [comment]
This PR is stuck since 5 days already:
- Was the origin of the nan identified (no comment received about it till now)?
  - If YES: can it be fixed?
  - If NOT: please continue to investigate (bugs should be fixed, rather than just hiding them under the carpet), but in the meanwhile we can just limit ourselves to keep track of it, for monitoring and debug purpose. Which is the frequency of the issue that you experienced so far? This will suggest you whether to add a LogError or a LogWarning
  - If any YES or NO, if you decide to mantain the protection as it is coded now, a LogError should be added anyhow, to take into account possible other occurrencies of the same.

**Comment by andrzejnovak** [comment]
> This PR is stuck since 5 days already:
> 
> * Was the origin of the nan identified (no comment received about it till now)?

No 

>   * If YES: can it be fixed?
>   * If NOT: please continue to investigate (bugs should be fixed, rather than just hidden under the carpet), but in the meanwhile we can just limit ourselves to keep track of it, for monitoring and debug purpose. Which is the frequency of the issue that you experienced so far? This will suggest you whether to add a LogError or a LogWarning

Frquency is sparse, there were only a few files with one or two events flagged, should this be LogError or LogWarning?

>   * If any YES or NO, if you decide to mantain the protection as it is coded now, a LogError should be added anyhow, to take into account possible other occurrencies of the same.

**Comment by perrotta** [comment]
> Frequency is sparse, there were only a few files with one or two events flagged, should this be LogError or LogWarning?

I would say that a LogError is more appropriate here, so that it won't be forgotten. But the LogWarning you already committed can work as well.

On the other hand: what the debug of (one or two of) those events pointed out? From where the nan's were generated in the debugged events?

**Comment by andrzejnovak** [comment]
@perrotta I don't know what to origin of the Nans/Infs is, They are values stored on the jet daughters and to dig into it further one would have to dig into how these are constructed. The jet itself doesn't look pathological, has defintie pT/mass/eta and number of non-problematic daughters.

**Comment by perrotta** [comment]
> @perrotta I don't know what to origin of the Nans/Infs is, They are values stored on the jet daughters and to dig into it further one would have to dig into how these are constructed. 

Please do!
You know now which are the exact events and jets to look at: it shouldn't be too hard to debug and retrieve the origin of the nan's. Depending on the outcome of the debug one will decide about the possible fix.

> The jet itself doesn't look pathological, has defintie pT/mass/eta and number of non-problematic daughters.

**Comment by andrzejnovak** [comment]
@perrotta  The jet->print() of the faulty jet is below. I don't think that Jet Reconstruction is the scope of this fix, maybe an issue should be opened and the responsible maintainers flagged.

To reproduce the issue:
`eventsToProcess = cms.untracked.VEventRange('303832:918:1081102780-303832:918:max')`
`process.source.fileNames = cms.untracked.vstring(
'/store/data/Run2017E/DoubleEG/MINIAOD/31Mar2018-v1/00000/5A8BC921-F437-E811-81ED-1866DAEB3370.root')`

```
Jet p/px/py/pz/pt: 241594/-58412.8/-57532.8/-227257/81988.3
    eta/phi: -1.74371/-2.36378
    # of constituents: 21
    Constituents:
      #0 p/pt/eta/phi: inf/inf/-1.74377/-2.36374    productId/index: 4:1022/1614
      #1 p/pt/eta/phi: 1.27675/0.361084/-1.93548/-1.63531    productId/index: 4:1022/425
      #2 p/pt/eta/phi: 0.646707/0.254395/-1.58501/-1.67104    productId/index: 4:1022/421
      #3 p/pt/eta/phi: 5.19222/1.5166/-1.90179/-2.39022    productId/index: 4:1022/1264
      #4 p/pt/eta/phi: 6.70202/1.30957/-2.31617/-2.82929    productId/index: 4:1022/1270
      #5 p/pt/eta/phi: 1.7118/1.04395/-1.07816/-2.64558    productId/index: 4:1022/1232
      #6 p/pt/eta/phi: 2.32983/0.834961/-1.68554/-2.21188    productId/index: 4:1022/1262
      #7 p/pt/eta/phi: 3.47681/0.729492/-2.24348/-2.7624    productId/index: 4:1022/1268
      #8 p/pt/eta/phi: 1.68492/0.713379/-1.50444/-2.17048    productId/index: 4:1022/1259
      #9 p/pt/eta/phi: 1.27904/0.685059/-1.23655/-2.38464    productId/index: 4:1022/1242
      #10 p/pt/eta/phi: 2.79286/0.658203/-2.12427/-2.78465    productId/index: 4:1022/1267
      #11 p/pt/eta/phi: 1.5161/0.602539/-1.57384/-2.17369    productId/index: 4:1022/1260
      #12 p/pt/eta/phi: 0.42834/0.271729/-1.0278/-2.68143    productId/index: 4:1022/1231
      #13 p/pt/eta/phi: 0.425851/0.270264/-1.02725/-2.47078    productId/index: 4:1022/1230
      #14 p/pt/eta/phi: 0.459339/0.260254/-1.16916/-2.90585    productId/index: 4:1022/1239
      #15 p/pt/eta/phi: 0.505971/0.251709/-1.32279/-2.87118    productId/index: 4:1022/1246
      #16 p/pt/eta/phi: 0.517942/0.246338/-1.37425/-2.94022    productId/index: 4:1022/1251
      #17 p/pt/eta/phi: 0.383544/0.202393/-1.25413/-2.52244    productId/index: 4:1022/1244
      #18 p/pt/eta/phi: 0.388701/0.187744/-1.35667/-2.4534    productId/index: 4:1022/1250
      #19 p/pt/eta/phi: 0.419694/0.182129/-1.47716/-1.84244    productId/index: 4:1022/1258
      #20 p/pt/eta/phi: 0.401061/0.176758/-1.45994/-2.50496    productId/index: 4:1022/1257
```

**Comment by perrotta** [comment]
@andrzejnovak, I'm trying to reproduce but if I run a "normal" nanoAOD production job on all events as in https://github.com/cms-sw/cmssw/pull/27238#issuecomment-505054094 I am not able to reproduce any crash.

Can you please give me the exact recipe to reproduce?

**Comment by andrzejnovak** [comment]
@perrotta Maybe it's due to the bypass that was implemented in Nano? You can try running https://github.com/cms-sw/cmssw/blob/master/RecoBTag/TensorFlow/test/test_deep_doublex_cfg.py but modified as in my last comment https://gist.github.com/andrzejnovak/b23ccc6fd4a119388b3b64587793c36c

**Comment by perrotta** [comment]
@rappoccio @knash @gkasieczka 

In 
```
process.source.eventsToProcess = cms.untracked.VEventRange('303832:918:1081148408-303832:918:1081148408')
process.source.fileNames = filesRelValTTbarPileUpMINIAODSIM
process.source.fileNames = cms.untracked.vstring( '/store/data/Run2017E/DoubleEG/MINIAOD/31Mar2018-v1/00000/5A8BC921-F437-E811-81ED-1866DAEB3370.root')
```
there is one jet of 82 TeV pt (!) whose first daughter has nan for p and pt (maybe they are just above some threshold?). See https://github.com/cms-sw/cmssw/pull/27238#issuecomment-505054094 for a log and https://github.com/cms-sw/cmssw/pull/27238#issuecomment-505643478 to reproduce (but also a nanoAOD job running on the same miniAOD output will hit the same).

I think that for you can be easier than for me (I'm kind of stuck in trying to identify where those crazy values were computed) or for the author of this band-aid PR to debug and check where those crazy values come from. Then, a fix should be applied.

Please let us know if this is a check that you can do quickly. Otherwise we will accept this PR as such, just to let the nanoAOD productions proceed: but the jets will remain bugged, and then a github issue will need to be opened, and tried to get solved quickly anyhow.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- The fix provides a rescue from nan/inf values in daughter jets parameters, and will allow nanoAOD and analysis to continue running even in presence of such pathological cases
- A fix for the wrong jet parameters has to be provided independently: issue #27374 has been opened for that
- Jenkins tests pass and show no differences

**Comment by fabiocos** [comment]
+1

**Comment by perrotta** [comment]
@andrzejnovak @ferencek : is a backport of this PR in 10_6 (or event 10_2) expected? nanoAOD production in those releasee should profit of it.

**Comment by andrzejnovak** [comment]
@perrotta Could be done, but v5 is ~finished right? What release is v6 gonna be in?

**Comment by perrotta** [comment]
> @perrotta Could be done, but v5 is ~finished right? What release is v6 gonna be in?

@andrzejnovak I don't know about the plans for nanoAOD releases, sorry. However 10_6 is the release for the legacy Run2 analyses, therefore I suppose that ALL forthcoming Run2 related nanoAOD productions will be processed with 10_6

**Comment by andrzejnovak** [comment]
@perrotta Wait, I have backported this right away then and it looks like it's merged #27328 #27329

**Comment by perrotta** [comment]
> @perrotta Wait, I have backported this right away then and it looks like it's merged #27328 #27329

Indeed... Sorry I overlooked it.
Thank you for remainding me.

**Pull request merged**